[18기 정승옥] 카카오 지도 마커 중복 오류 개선

luxcaeli opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
- 지도, 필터 컴포넌트 레이아웃 작업 및 일부 기능 구현 진행 
		 
## 수정 사항들 자세한 내용
- 지도, 필터 컴포넌트 레이아웃 작업 완료
- 가격 범위, 업종 카테고리 선택 기능 구현
- 카카오 클러스터링, 마커 지도 표시, 지도 검색 api 활용

## 기타 질문 및 특이 사항
- 카테고리(업종별)를 선택할 때 if-else문으로 조건을 나누어 usestate를 활용했는데 리팩토링해야 되는 부분 알려주시면 감사하겠습니다!
- 가격 범위를 양쪽에서 설정하는 예제를 못 찾아서 현재 라이브러리를 사용했는데 참고할만한 예제가 있으면 추천 부탁드립니다!

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.
- [x] Git rebase와 squash를 했고 커밋 수가 하나 인것을 확인 했습니다.

bonbon0202 commented on src/Pages/Main/Components/FoodMap/Filter/PriceOption/PriceOption.js:
```diff
@@ -0,0 +1,75 @@
+  markLabel: { fontSize: '10px', top: '13px' },
+  track: { height: 10, background: '#00b8ff' },
+  rail: { height: 10, background: '#03a9f4' },
+})(Slider);
```
라이브러리 잘 찾아서 이용하셨네요! 
넵 range multiple handle구현이 쉽지 않으셨을 것 같습니다.
프로젝트 후 시간되실 때 도전해보시길 바라요 ! :) 
바닐라 자바스크립트로 구현된 참고 예제를 드릴게요! 
(https://www.youtube.com/watch?v=pTlsnFLiK6c)

bonbon0202 commented on src/Pages/Main/Components/FoodMap/Filter/PriceOption/PriceOption.js:
```diff
@@ -0,0 +1,75 @@
+  const handleChange = (e, newValue) =>
+    newValue[0] !== newValue[1] && setValue(newValue);
+
+  let title = '';
+  if (value[0] === 0 && value[1] === 20000) title = '전체';
+  else if (value[0] === 0)
+    title = `${Number(value[1]).toLocaleString()}원 이하`;
+  else if (value[1] === 20000)
+    title = `${Number(value[0]).toLocaleString()}원 이상`;
+  else
+    title = `${Number(value[0]).toLocaleString()}원 ~ ${Number(
+      value[1]
+    ).toLocaleString()}원`;
```
승옥님! :) 
조건문 잘 작성해주셨습니다 👏🏻
useEffect를 사용하셔도 좋겠네요!

luxcaeli replied:
```jsx
  useEffect(() => {
    if (value[0] === 0 && value[1] === 20000) setTitle('전체');
    else if (value[0] === 0)
      setTitle(`${Number(value[1]).toLocaleString()}원 이하`);
    else if (value[1] === 20000)
      setTitle(`${Number(value[0]).toLocaleString()}원 이상`);
    else
      setTitle(
        `${Number(value[0]).toLocaleString()}원 ~ ${Number(
          value[1]
        ).toLocaleString()}원`
      );
  }, [value]);
```

신영님! 리뷰 보고 위에 코드로 수정했는데 조건에 따라 usestate를 여러번 선언하는거 같아서 더 좋은 방법 없을까요 ㅠㅠ

bonbon0202 replied:
승옥님! :) 해당 경우는 조건이 4개이고 그에 따라서 결과값이 4개로 나오는 경우이기 때문에 위와 같이 작성하시면 될 것 같습니다. 
다만, 가독성을 고려 하신다면 아래와 같이 수정하실 수도 있습니다. 
```js
  useEffect(() => {
    if (value[0] === 0 && value[1] === 20000) return setTitle('전체');

    if (value[0] === 0) return setTitle(`${Number(value[1]).toLocaleString()}원 이하`);

    if (value[1] === 20000) return setTitle(`${Number(value[0]).toLocaleString()}원 이상`);

    setTitle(`${Number(value[0]).toLocaleString()}원 ~ ${Number(value[1]).toLocaleString()}원`);
  }, [value]);
```

luxcaeli replied:
각각 if 문 선언하는게 가독성이 좋아보이네요 수정했습니다!

bonbon0202 commented on src/Pages/Main/Components/FoodMap/Filter/StoreOption/StoreOption.js:
```diff
@@ -0,0 +1,88 @@
+      }
+    }
+  };
+  const checkCount = selectCategory.filter(Boolean).length;
+  const countType = [selectCategory.length - 1, checkCount];
```
상수는 useState아래에서 함께 사용해주시면 가독성에 더 좋을 것 같습니다 :)

luxcaeli replied:
```jsx
const [checkCount, setCheckCount] = useState(0);

useEffect(() => {
    setCheckCount(selectCategory.filter(Boolean).length);
    getCategoryCheck(selectCategory);
  }, [selectCategory]);
```

카테고리 선택을 관리하는 배열의 값이 변경될때 usestate를 사용해서 카운트 값이 변하도록 수정했습니다!

bonbon0202 commented on src/Pages/Main/Components/FoodMap/Filter/StoreOption/StoreOption.js:
```diff
@@ -0,0 +1,88 @@
+      dataset: { target },
+    } = e.target;
+
+    const saveType = [
```
select category를 배열로 관리하시는 방법도 있고 객체로 관리하시는 방법도 가능할 것 같습니다 :)

luxcaeli replied:
신영님! 혹시 배열로 관리할때랑 객체로 관리했을때랑 어떤 차이점이 있을지 추가 설명 부탁드려도 될까요 ㅠㅠ 감이 잘 안 오네요

bonbon0202 replied:
여러 방법이 있다는 정도 알려드리는 것이었습니다 :) 
배열로 관리하시면 인덱스에 의존해야하지만 객체로 관리할 경우에는 보다 더 명확한 키 값에 의존하게 된다는 차이점이 있겠네요!

bonbon0202 commented on src/Pages/Main/Components/FoodMap/Filter/StoreOption/StoreOption.js:
```diff
@@ -0,0 +1,88 @@
+      ),
+    ];
+
+    if (Number(target) === 1) {
+      setSelectCategory(saveType[0]);
+    } else {
+      if (
+        selectCategory.filter(Boolean).length === selectCategory.length - 2 &&
+        selectCategory[Number(target) - 1] === false
+      ) {
+        setSelectCategory(saveType[0]);
+      } else {
+        setSelectCategory(saveType[1]);
+      }
+    }
```
가독성에 따라 다르겠지만, 
해당 조건문을 삼항연산자와 || 를 이용해서 반복되는 구문을 줄여주고 가독성을 좀 더 높일 수 있을 것 같습니다.
조건문은 아래와 같이 상수에 선언해 관리하실 수 있습니다.
```suggestion
    const targetCondition = Number(target) === 1
```

luxcaeli replied:
```jsx
const handleCategorySubOrAll = (target, saveType) => {
    const nowCheck = selectCategory.filter(Boolean).length;
    const nowSub = selectCategory[Number(target - 1)];

    return nowCheck === selectCategory.length - 2 && !nowSub
      ? setSelectCategory(saveType[0])
      : setSelectCategory(saveType[1]);
  };

const checkCategory = (e, target) => {
    // 전첵 선택, 카테고리 선택 2가지 경우
    const saveType = [
      selectCategory.map((select, index) => (index === 0 ? !select : false)),
      selectCategory.map((select, index) =>
        index === 0 ? false : index + 1 === Number(target) ? !select : select
      ),
    ];
    const targetCondition = Number(target);
    targetCondition === 1
      ? setSelectCategory(saveType[0])
      : handleCategorySubOrAll(target, saveType);
  };
```

삼항연산자와 || 모두 사용하는 방법은 구현방법이 생각이 잘 안나서 일단 삼항연산자 활용과 else 부분 조건들을 함수로 담아서 수정해봤습니다! 삼항연산자와 || 모두 사용하는 방법은 어떤 식으로 구현해야할까요 ㅠㅠ

bonbon0202 replied:
첫번째 조건과 두번째 조건의 결과 값이 같으므로 아래와 같이 작성할 수 있겠네요! :) 

```js
    const targetCondition = Number(target) === 1;
    const condition =
      selectCategory.filter(Boolean).length === selectCategory.length - 2 &&
      selectCategory[Number(target) - 1] === false;

    setSelectCategory(targetCondition || condition ? saveType[0] : saveType[1]);
```

luxcaeli replied:
ContextAPI 사용하면서 코드가 변경돼서 다시 수정해보겠습니다!

bonbon0202 commented on src/Pages/Main/Components/FoodMap/FoodMap.js:
```diff
@@ -0,0 +1,103 @@
+};
+
+const Map = styled.div`
+  width: 100%;
```
필요한 속성일까요?!

luxcaeli replied:
앗 ㅠㅠ 수정하겠습니다!

bonbon0202 reviewed (commented):
승옥님! 수고하셨습니다 :) 
리뷰에 답변 드렸습니다!! 참고하셔서 리팩토링 해보세요!

bonbon0202 commented on src/Pages/Main/Components/FoodMap/Filter/StoreOption/StoreOption.js:
```diff
@@ -0,0 +1,238 @@
+  const { checkCategory } = action;
+
+  const toggleSectionList = () => {
+    setSectionToggle(
+      sectionToggleCheck === 0
+        ? Array(sectionCounts).fill(true)
+        : Array(sectionCounts).fill(false)
+    );
```
조건문을 어디에 작성하느냐에 따라 불필요한 반복되는 코드를 줄일 수 있어 보입니다 :)

luxcaeli replied:
```jsx
const toggleSectionList = () => {
    const toggleCondition = sectionToggleCheck === 0 ? true : false;
    setSectionToggle(Array(sectionCounts).fill(toggleCondition));
  };
```
넵! 수정했습니다!

bonbon0202 commented on src/Pages/Main/Components/FoodMap/FoodMap.js:
```diff
@@ -0,0 +1,153 @@
+        storeImage.src = imageUrl;
+        storeImage.alt = '가게 이미지';
+        sliderBox.append(storeImage);
+      });
+      overlayBox.append(sliderBox);
+
+      const prevBtn = document.createElement('button');
+      const nextBtn = document.createElement('button');
+      const storeDetail = document.createElement('div');
+      prevBtn.className = 'prev';
+      nextBtn.className = 'next';
+      storeDetail.className = 'storeDetail';
+      storeDetail.innerHTML = `
+        <h2 class="title">${data.name}</h2>
+        <p class="detail">
+          <span class="rate">${data.rate}</span>
+          <span class="type">${data.type}</span>
+        </p>
+      `;
+      overlayBox.append(storeDetail, prevBtn, nextBtn);
+
+      const handleSlider = e => {
+        const btnType = e.target.className;
+        const imageUrls = data.url;
+        const moveSize = sliderBox.offsetWidth / imageUrls.length;
+        let leftValue = Number(sliderBox.style.left.split('px')[0]);
+
+        if (btnType === 'prev') {
+          leftValue += moveSize;
+          leftValue <= 0 && (sliderBox.style.left = `${leftValue}px`);
+        } else {
+          leftValue -= moveSize;
+          leftValue > sliderBox.offsetWidth * -1 &&
+            (sliderBox.style.left = `${leftValue}px`);
+        }
+      };
+
+      prevBtn.addEventListener('click', handleSlider);
+      nextBtn.addEventListener('click', handleSlider);
+
```
리액트에서 dom에 접근해 사용해야할 떄가 있습니다.
그럴때 리엑트에서는 ref라는 것을 사용하고, 함수형 컴포넌트에서는 useRef 라는 hook 함수를 사용합니다.
아래의 문서들을 참조해보세요! 
공식문서 : https://ko.reactjs.org/docs/hooks-reference.html#useref
hooks로 carousel 만들기 : https://krpeppermint100.medium.com/js-react-hooks%EB%A1%9C-carousel-slider-%EB%A7%8C%EB%93%A4%EA%B8%B0-2e558151bbee

luxcaeli replied:
```jsx
const overlay = new kakao.maps.CustomOverlay({
        content: overlayBox,
        clickable: true,
        position: mark.getPosition(),
        xAnchor: 0.5,
        yAnchor: 1.3,
        zIndex: 3,
      });
```
마커 오버레이 컴포넌트로 만들고 content에 담으면 리액트 JSX에 있는 이벤트들이 없어지는 문제가 생겨 직접 돔을 생성해서 했습니다 ㅠㅠ 
지금 생각한 방법은 오버레이 생성부분을 함수로 따로 js 파일에 담는 방법을 생각중인데 혹시 컴포넌트를 사용해도 이벤트가 사라지지 않는 방법이 있을까요????

we-dohyeon commented:
승옥님! conflict 해결하시고 다시 올려주세요 :)

Commit:
[18기 정승옥] Conflict 해결중

[18기 정승옥] 지도 작업중

[18기 정승옥] 지도 작업중

[18기 정승옥] 카카오 지도 마커 중복 오류 개선

[18기 정승옥] 마커 중복 코드 수정

[18기 정승옥] 지도 마커 중복 해결 작업중

joonsikyang commented:
@Jeong-seungok 승옥님 다른 PR merge 되는 과정에서 다시 conflict 발생하였습니다 한 번 더 해결 부탁드립니다-!
